drivers/gpio/gpio_mcp23x17.c

```diff
@@ -1,13 +1,18 @@
 /*
  * Copyright (c) 2020 Geanix ApS
+ * Copyright (c) 2020 Prevas A/S (I2C added)
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#if defined(CONFIG_MCP23X17_DEV_TYPE_I2C)
+#define DT_DRV_COMPAT microchip_mcp23017
+#elif defined(CONFIG_MCP23X17_DEV_TYPE_SPI)
 #define DT_DRV_COMPAT microchip_mcp23s17
+#endif
```

Comment:
This change means that either mcp23017 or mcp23s17 is supported, but not both. That shouldn't be the case: the driver should be able to support both types of devices at the same time.

I suggest doing something like this at the bottom:

```
#define DT_DRV_COMPAT microchip_mcp23017
#define MCP23017_DEFINE(inst) ... make the struct device for an mcp23017 instance 'inst' ...
DT_INST_FOREACH_STATUS_OKAY(MCP23017_DEFINE)

#undef DT_DRV_COMPAT
#define MCP23S17_DEFINE(inst) ... make the struct device for an mcp23s17 instance 'inst' ...
DT_INST_FOREACH_STATUS_OKAY(MCP23S17_DEFINE)
```

Each device's configuration structure should be able to store any device-type-specific information that the functions above might need, like a pointer to a bus-specific I/O structure as is done in the BME280 driver mentioned by @pabigot in his review.